### PR TITLE
Add install/uninstall integration test

### PR DESCRIPTION
**Pull request by Dean-Coakley**
### Proposed changes

* Convert install/uninstall to be an integration test
* Run install/uninstall test on push CI via `make integration-test` target
* Test no longer depends on `nginx` running
* Install/uninstall test now uses local package
* Test run against docker container. `ubuntu` is the only supported image at the moment.

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [`CONTRIBUTING`](https://github.com/nginx/agent/blob/main/docs/CONTRIBUTING.md) document
- [x] If applicable, I have added tests that prove my fix is effective or that my feature works
- [x] If applicable, I have checked that any relevant tests pass after adding my changes
- [x] I have updated any relevant documentation ([`README.md`](https://github.com/nginx/agent/blob/main/README.md))

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -165,8 +166,13 @@ build-benchmark-docker: ## Build benchmark test docker image for NGINX Plus, nee
 		-f test/docker/Dockerfile .
 
 # Install tests
-test-install: ## Run agent install test
-	GOWORK=off CGO_ENABLED=0 go test -v ./test/install
+test-install: ## Run agent install/uninstall test for release
```
Comment (oliveromahony) [inline Makefile:169]: this should probably roll up into an integration target, with multiple tests suites running against that and a new CI/CD stage

File: Makefile [hunk 2]
```diff
@@ -165,8 +166,13 @@ build-benchmark-docker: ## Build benchmark test docker image for NGINX Plus, nee
...
+	AGENT_PACKAGE_FILE=${shell find ./build/packages/deb/*${OS_CODENAME}_${LOCAL_ARCH}.deb -type f -exec realpath {} \;} \
```
Comment (craigell) [inline Makefile:170]: Think we should let the pipeline dictate where the AGENT_PACKAGE_FILE variable is?

File: .github/workflows/release-branch.yml [hunk 3]
```diff
@@ -151,6 +151,8 @@ jobs:
         run: |
           echo "$INDIGO_GPG_AGENT" | base64 --decode > .key.asc
           make clean package
+      - name: Test install and uninstall
```
Comment (dhurley) [inline .github/workflows/release-branch.yml:154]: This only needs to be added to the ci.yml file as that workflow will trigger the release.yml workflow

File: Makefile [hunk 4]
```diff
@@ -180,11 +180,11 @@ performance-test: ## Run performance tests
 	$(CONTAINER_CLITOOL) run -v ${PWD}:/home/nginx/$(CONTAINER_VOLUME_FLAGS) --rm nginx-agent-benchmark:1.0.0
 
 integration-test: local-deb-package
-	PACKAGE=${PACKAGE_NAME} BASE_IMAGE=${BASE_IMAGE} go test ./test/integration/api
+	PACKAGE_NAME=${PACKAGE_NAME} BASE_IMAGE=${BASE_IMAGE} go test ./test/integration/install
+	PACKAGE_NAME=${PACKAGE_NAME} BASE_IMAGE=${BASE_IMAGE} go test ./test/integration/api
 
 test-bench: ## Run benchmark tests
 	cd test/performance && GOWORK=off CGO_ENABLED=0 go test -mod=vendor -count 5 -timeout 2m -bench=. -benchmem metrics_test.go
-	cd test/performance && GOWORK=off CGO_ENABLED=0 go test -mod=vendor -count 1 -bench=. -benchmem user_workflow_test.go
```
Comment (oliveromahony) [inline Makefile:187]: why was this removed?

Comment (Dean-Coakley) [reply Makefile:187]: Accidental. I think the install test used to have a similar filename a while back. 

Thanks for catching this. Reverting.

File: test/integration/install/install_uninstall_test.go [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+const (
+	osReleasePath  = "/etc/os-release"
+	maxFileSize    = int64(20000000)
+	maxInstallTime = 30 * time.Second
+)
+
+var (
+	AGENT_PACKAGE_FILENAME = os.Getenv("PACKAGE_NAME")
+	AGENT_PACKAGE_FILEPATH = "/agent/build/" + AGENT_PACKAGE_FILENAME + ".deb"
```
Comment (craigell) [inline test/integration/install/install_uninstall_test.go:26]: Should we hardcode the package extension? This might not always run on a Debian or Ubuntu environment.

File: test/docker/integration/install/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,27 @@
+FROM ubuntu:22.04 as install
```
Comment (oliveromahony) [inline test/docker/integration/install/Dockerfile:1]: can we reuse the OSS docker files and add in multi-stages, one for the operating system and NGINX. The second for the agent. You can target specific stages in the docker file through the cli: see the options` --stage ` https://docs.docker.com/engine/reference/commandline/build/#options and this can be done in compose in a similar way: https://stackoverflow.com/a/53101932/515970

File: test/integration/install/install_uninstall_test.go (hunk 5 above)
Comment (Dean-Coakley) [reply test/integration/install/install_uninstall_test.go]: I should've stated in this PR that the scope was solely ubuntu as a starting point.

However this was easy to support so I decided to include in: https://github.com/nginx/agent/pull/147/commits/bb3607a7d4d2386822297270586274f800e8b4f7

**Review by aphralG — APPROVED** [review]

File: test/integration/install/install_uninstall_test.go [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+		"AgentDynamicConfigFile": "/etc/nginx-agent/agent-dynamic.conf",
+	}
+
+	agentPackageFilePath := getPackagePath(string(osReleaseContent))
+
+	// Check the agent package is present
+	agentPkg, err := agentContainer.CopyFileFromContainer(context.Background(), agentPackageFilePath)
+	assert.NoError(t, err)
+
+	a, err := io.ReadAll(agentPkg)
```
Comment (dhurley) [inline test/integration/install/install_uninstall_test.go:91]: Could you give these variables better names instead of using single letters?

**Review by dhurley — COMMENTED** [review]
Looks good just have a few minor comments

File: test/integration/install/install_uninstall_test.go [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+	assert.NoError(t, err)
+
+	f, err := os.CreateTemp("", "tmpfile-")
```
Comment (oliveromahony) [inline test/integration/install/install_uninstall_test.go:94]: I don't get the need for lines 94-102? Should you be reading the package location and verifying that instead?

File: test/integration/install/install_uninstall_test.go (hunk 7 above)
Comment (oliveromahony) [inline test/integration/install/install_uninstall_test.go:91]: Can you add a more descriptive variable name? Something that helps the code reader with the context

**Review by dhurley — APPROVED** [review]

**Review by oliveromahony — APPROVED** [review]
